### ISSUE TITLE
Node stops producing blocks after transaction from new accounts

### ISSUE THREAD

**Issue by igamigo** [0xPolygonMiden/miden-node#211]
This is related to https://github.com/0xPolygonMiden/miden-node/issues/176.

After #200 was merged, submitting proofs for new accounts works fine on the client, which means the RPC does not return an error:
```
Running `target/release/miden-client tx new consume-notes 0x444bca3cbb5d302d 0xbd8f70675194683eec7b8f8d5b29c7306970dcaaca44726dac9c07624fafa2ae`
2024-02-05T18:03:34.138529Z  INFO miden_client::cli::transactions: Executed transaction, proving and then submitting...
2024-02-05T18:03:34.190356Z  INFO Proving program: miden_prover: Generated execution trace of 70 columns and 131072 steps (27% padded) in 51 ms
2024-02-05T18:03:35.077097Z  INFO miden_client::client::transactions: Proved transaction, submitting to the node...
2024-02-05T18:03:35.115166Z  INFO miden_client::store::transactions: Transaction id 0x134e33c69d0db99f744922a98b4a54edb85fc58862243f0c91afddafc0bcba85
2024-02-05T18:03:35.115174Z  INFO miden_client::store::transactions: Transaction account id: 0x444bca3cbb5d302d
```

However, after the proof is actually submitted, the node is then locked in a state where it cannot generate blocks, and fails periodically with the following error:

```
2024-02-05T18:05:09.573085Z ERROR block_producer:try_build_block:build_block: miden-block-producer: block-producer/src/block_builder/mod.rs:73: error: transaction batches and store contain different hashes for some accounts. Offending accounts: [AccountId(4921249380022693933)]

...

2024-02-05T18:10:19.571170Z ERROR block_producer:try_build_block:build_block: miden-block-producer: block-producer/src/block_builder/mod.rs:73: error: transaction batches and store contain different hashes for some accounts. Offending accounts: [AccountId(4921249380022693933)]
```

This does not happen when executing transactions against accounts that have already been added to the accounts tree (like the ones generated on genesis).
Seems like there might be a check on the block producer that does not take the newer changes from #200 into account.

**Comment by bobbinth** [comment]
> If you look further into the error that you copied you will see that the problem here is when we are building the `block_witness` we are making a certain number of checks and one of them is not accounting for new accounts. It is comparing the received account hash and the one in store and throwing an error if they are different. Obviously for a new account the store one would be Word 0 and the received one would be Word X, hence erroring out.

This is happening somewhere [here](https://github.com/0xPolygonMiden/miden-node/blob/main/block-producer/src/block_builder/prover/block_witness.rs#L217-L233), right? If so, the fix should be similar to what we did in #200.

> After fixing this issue I found another one that I am investigating at the moment where block proving doesn't work anymore because of the fixes to accept new accounts.

What kind of an error are you getting?

> I have also found multiple situations where the node arrives to a deadlock. Will need to investigate further in the future because there are definitely hidden bugs.

I think the main issue right now is that the node does not recover from "bad transactions". Basically, if some transaction causes issues with block building (for w/e reason), the node has no mechanism for discarding it. That is, it will just keep trying to add it to the block. The assumption here is that if a transaction made it into the transaction queue, it should not cause issues downstream - but in this case because of a separate bug, this is not the case. Let's create a separate issue to figure out how we want to deal with such situations.

**Comment by bobbinth** [comment]
> > but in this case because of a separate bug, this is not the case. Let's create a separate issue to figure out how we want to deal with such situations.
> 
> Could you be clearer I have not fully understood your point please.

One of the problems we have is that if there is a transaction in the transaction queue which causes block production to fail, the node will deadlock (it will try to create a block with the same transaction over and over again). In this case, we know how to fix the problem (i.e., the transaction in question should actually not fail), but it would be good to have a more general solution for such situations (i.e., to avoid deadlocks even if there are bugs).

**Comment by bobbinth** [comment]
> > What kind of an error are you getting?
> 
> `�[2m2024-02-05T12:46:00.031022Z�[0m �[31mERROR�[0m �[1mblock_producer�[0m:�[1mtry_build_block�[0m:�[1mbuild_block�[0m: �[2mmiden-block-producer�[0m�[2m:�[0m �[2mblock-producer/src/block_builder/mod.rs�[0m�[2m:�[0m�[2m73�[0m�[2m:�[0m �[1merror�[0m: failed to compute new block: program execution failed`
> 
> After submitting a transaction using new accounts no further blocks can be proven.

This is likely due to some assumption within the MASM file which we use to compute the new block header being no longer valid.

**Comment by phklive** [comment]
I believe after testing that this issue has been resolved. I have not seen it arise again.

These PR should have fixed it: 
- https://github.com/0xPolygonMiden/miden-node/pull/200
- https://github.com/0xPolygonMiden/miden-node/pull/230
- https://github.com/0xPolygonMiden/miden-node/pull/232

Opened this following issue for improved transaction handling and queue management as discussed with @bobbinth and @hackaugusto on the call Monday: https://github.com/0xPolygonMiden/miden-node/issues/235
